reyna/src/com/b2msolutions/reyna/Dispatcher.java

```diff
@@ -49,20 +58,34 @@ protected Result sendMessage(Message message, HttpPost httpPost, HttpClient http
         return this.tryToExecute(httpPost, httpClient);
     }
 
-    public static Result canSend(Context context) {
+    public static Result canSend(Context context) throws ParseException {
+        // *****************************
+        // TODO all the checks here should be combined with our window blackout feature
+        // *****************************
+
         ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
         NetworkInfo info = connectivityManager.getActiveNetworkInfo();
-        if(info == null || !info.isConnectedOrConnecting()) {
+        if (info == null || !info.isConnectedOrConnecting()) {
             return Result.NOTCONNECTED;
         }
+        int type = info.getType();
+        Preferences preferences = new Preferences(context);
+
+        BlackoutTime blackoutTime = new BlackoutTime(context);
+        if (canSendOnWlanCharging(context, type, preferences, blackoutTime)) return Result.BLACKOUT;
+        if (canSendOnWlanDischarging(context, type, preferences, blackoutTime)) return Result.BLACKOUT;
```

Comment:
Rather than always having those two property connections with Charging and Discharging, wouldn't it be easier to test separately if the power status and settings or the connection status and settings stop us from sending. There doesn't seem to be a reason to do that together.
Eg.
if (power.isCharging && !canSendOnCharging()) return Result.BLACKOUT;
if (!power.isCharging && !canSendOffCharge()) return Result.BLACKOUT;
if (type == ConnectivityManager.TYPE_WIFI && !blackoutTime.canSendOnWlan(new GregorianCalendar())) return Result.BLACKOUT;
if (type == ConnectivityManager.MOBILE && !blackoutTime.canSendOnWwan(new GregorianCalendar())) return Result.BLACKOUT
or something like that.